docs/REPRODUCIBILITY.md

```diff
@@ -0,0 +1,267 @@
+
+It is possible to mitigate, but not completely solve, the above issue by adding a specific version:
+
+```sh
+apk add bash=5.1.16-r2
+```
+
+The above does not completely solve it, as sometimes alpine packages change their bits _even for the same version_. Thus, running
+`apk add bash=5.1.16-r2` on the same base `alpine:3.16` can give different bits on different days. Alpine packages are changed "under the covers"
+when bugs are found or updates are provided.
+
+In order to resolve this issue, both the external/upstream `linuxkit/*` packages and the `lfedge/eve-*` packages in this repository follow the same principles.
+
+1. Create an alpine "base" image: `linuxkit/alpine` and `lfedge/eve-alpine`
+1. All packages included in the composed EVE image **must** use `FROM lfedge/eve-alpine` or, in the case of linuxkit packages, `FROM linuxkit/alpine`, in their `Dockerfile`. `FROM alpine` is strictly forbidden.
+1. The base images source from upstream alpine, download all necessary apk packages needed by downstream, `lfedge/eve-alpine` for EVE packages and `linuxkit/alpine` for linuxkit packages, and cache them.
+1. When downstream images need to install packages, they run `apk add package` which sources it solely from the cache in the base image.
+
+The base images - `lfedge/eve-alpine` and `linuxkit/alpine` - act as pinning cache, with precise versions and bits of each package.
+
+Only when the cache is regenerated, i.e. when `lfedge/eve-alpine` or `linuxkit/alpine` is rebuilt, can these change.
+
+Further, the caches maintain hashes of the local cache contents, to ensure that any changes are known and captured.
+
+To make this more concrete, we will walk through a specific example, `lfedge/eve-storage-init`, which is based in [../pkg/storage-init](../pkg/storage-init/).
+
+The Dockerfile is as follows:
+
+```dockerfile
+FROM lfedge/eve-alpine:145f062a40639b6c65efa36bed1c5614b873be52 as build
+ENV PKGS alpine-baselayout musl-utils bash glib squashfs-tools util-linux e2fsprogs e2fsprogs-extra keyutils dosfstools coreutils sgdisk smartmontools
+RUN eve-alpine-deploy.sh
+
+FROM scratch
+COPY --from=build /out/ /
+COPY storage-init.sh /
+
+WORKDIR /
+ENTRYPOINT []
+CMD ["/storage-init.sh"]
+```
+
+Note 3 key elements:
+
+* The final image is based on `scratch`
+* The base image for installing necessary packages is **not** `alpine:3.16`, but rather the eve alpine cache image `lfedge/eve-alpine:145f062a40639b6c65efa36bed1c5614b873be52`
+* The packages needed to be installed are: `alpine-baselayout musl-utils bash glib squashfs-tools util-linux e2fsprogs e2fsprogs-extra keyutils dosfstools coreutils sgdisk smartmontools`
+
+The usage of `PKGS` environment variable and the script `eve-alpine-deploy.sh` is a simplified way to install packages. It is the equivalent of calling:
+`apk add alpine-baselayout musl-utils bash glib squashfs-tools util-linux e2fsprogs e2fsprogs-extra keyutils dosfstools coreutils sgdisk smartmontools`, but
+do so from the cache.
```

Comment:
Ok so since the base image is always fixed to a hash, these packages installed are guaranteed to be same every time right ?

Reply:
Yes. As long as you use the same hash-based version of the base image, then `apk add` will use the fixed ones in the cache, which is unchanged.